### DISCUSSION TITLE
Guidance for building servers that act as proxy clients to many servers

### DISCUSSION THREAD

**Discussion by ggoodman** [modelcontextprotocol/modelcontextprotocol#94] [Ideas - General]
### Pre-submission Checklist

- [X] I have verified this would not be more appropriate as a feature request in a specific repository
- [X] I have searched existing discussions to avoid duplicates

### Your Idea

I think there is lots of potential for an MCP server to act as a single point of configuration for many other MCP servers. This would benefit apps looking to act as MCP clients as they could thereby offload all of this management logic to an MCP proxy app.

Here is a straw man for a CLI experience:
- `mcp setup claude` -- Install `mcp run --scope claude` as a binary in `claude.json`.
- `mcp server search github` -- Search the list of known MCP servers for ones that have the keyword, `GitHub`.
- `mcp server install --scope claude @namespace/mcp-server` -- Install an MCP server locally, scoped to `claude`.
- `mcp server install @namespace2/mcp-server-2` -- Install _another_ MCP server locally for all scopes (clients).
- `mcp run --scope <scope>` -- The entrypoint for the binary to act as a `stdio` MCP server.

> [!NOTE]
> In a world where some form of local service discovery is added to the protocol (or becomes a community convention), step 1 might even be skipped. It would be enough to install the binary and have it start on login and listen in whatever way works with local service discovery.

This sort of UX would allow many clients to benefit from a single local experience for configuring LLM context providers. It would be a perfect place for users to authenticate themselves to 3rd party services without every MCP client needing to re-implement this logic over and over. It might even allow the protocol to remain agnostic to this sort of bearer credential. The protocol could limit itself to the mutual authentication between client and server.

**Spec changes**

To help support this sort of hypothetical setup, I think the spec might benefit from some extra level of namespacing. Here are some area that I can think of based on the parts of the spec I've interacted with:

1. Server [InitializeResult](https://github.com/modelcontextprotocol/specification/blob/bb5fdd282a4d0793822a569f573ebc36804d38f8/schema/schema.ts#L160-L176) -- This assumes that: 1) the Server is a single, homogenous entity; 2) that a single set of instructions can govern the entire server; and 3) that a single version of the protocol is adequate. Of these, I think (2) is the thing that is most likely going to be problematic. The server _could_ act as a homogenous veneer over a heterogeneous set of other servers and _could_ expose a single protocol version to its clients.
2. Named entities such as [Prompt](https://github.com/modelcontextprotocol/specification/blob/bb5fdd282a4d0793822a569f573ebc36804d38f8/schema/schema.ts#L554-L570) -- The protocol assumes that named entities live in a collision-free namespace. In an MCP proxy world, Tools, Prompts, and any other 'named' entities could have conflicting names between servers. This is also true today where multiple servers are free to declare the same tool. It's not clear how a client would resolve ambiguity in such a situation. A server could transparently add some namespacing or the protocol could be extended with some form of support.
3. `list_changed` [notifications](https://github.com/modelcontextprotocol/specification/blob/bb5fdd282a4d0793822a569f573ebc36804d38f8/schema/schema.ts#L673-L678) -- These notifications would invalidate the whole namespace when finer-grained invalidation might be more efficient. If a namespacing concept were to be introduced, these notifications could be augmented to limit their applicability to zero or more namespaces.

As I haven't implemented anywhere close to 100% of the spec, the above is almost certainly not an exhaustive list. But hopefully it offers some inspiration for some potential future directions in the spec.

### Scope

- [X] Protocol Specification
- [ ] SDK Features
- [ ] Documentation
- [ ] Developer Experience
- [ ] Other

**Comment by jspahrsummers** [comment]
Thanks for flagging this! These are great points to think about.

Regarding your specific list:

> 1. Server [InitializeResult](https://github.com/modelcontextprotocol/specification/blob/bb5fdd282a4d0793822a569f573ebc36804d38f8/schema/schema.ts#L160-L176) -- This assumes that: 1) the Server is a single, homogenous entity; 2) that a single set of instructions can govern the entire server; and 3) that a single version of the protocol is adequate. Of these, I think (2) is the thing that is most likely going to be problematic. The server could act as a homogenous veneer over a heterogeneous set of other servers and could expose a single protocol version to its clients.

Can you share more about the concern you think exists with (2)? If "instructions" refer to "how to use all the tools that the server provides" (for example), a server-of-servers seems like it could get most of the way by just naively concatenating the instructions from all its child servers. I'm curious if that seems inadequate, though!

> 2. Named entities such as [Prompt](https://github.com/modelcontextprotocol/specification/blob/bb5fdd282a4d0793822a569f573ebc36804d38f8/schema/schema.ts#L554-L570) -- The protocol assumes that named entities live in a collision-free namespace. In an MCP proxy world, Tools, Prompts, and any other 'named' entities could have conflicting names between servers. This is also true today where multiple servers are free to declare the same tool. It's not clear how a client would resolve ambiguity in such a situation. A server could transparently add some namespacing or the protocol could be extended with some form of support.

I think it's _probably_ reasonable for clients to automatically namespace everything per-server (and this could apply to the server-of-servers with its children servers as well). There might be problems with this I'm not seeing, though. We certainly need to give more thought here.

> 3. list_changed [notifications](https://github.com/modelcontextprotocol/specification/blob/bb5fdd282a4d0793822a569f573ebc36804d38f8/schema/schema.ts#L673-L678) -- These notifications would invalidate the whole namespace when finer-grained invalidation might be more efficient. If a namespacing concept were to be introduced, these notifications could be augmented to limit their applicability to zero or more namespaces.

We probably want to extend these notifications to support delta updates regardless ("these things were added, these were deleted"), which might solve this problem too.

**Reply by richardkmichael** [reply]
With respect to `instructions` and concatenation, perhaps it could change to being an object (instead of a string), and then it could gain `listChanged`?  Using objects seems to have nicely provided a point for future extensions.  FWIW, I have tried to use `instructions` in my MCP servers, with limited success due to lack of client awareness (I think); hopefully this improves.

**Comment by ggoodman** [comment]
> Can you share more about the concern you think exists with (2)? If "instructions" refer to "how to use all the tools that the server provides" (for example), a server-of-servers seems like it could get most of the way by just naively concatenating the instructions from all its child servers. I'm curious if that seems inadequate, though!

@jspahrsummers, I think it's not adequate because I envision the set of configured servers changing over time. Even within the life-cycle of a single session. Maybe the tool allows servers to be _dynamically installed_? In this case, the instructions will be invalidated and the server will have no mechanism to inform clients.

**Comment by ondrsh** [comment]
Servers acting as aggregators can work with the existing protocol by implementing their own namespacing. Some aggregators may not even want to expose their child servers directly; they can combine multiple low-level servers (e.g., filesystem, database) into a high-level, domain-specific API. The idea of having a server with the main purpose of reducing repeated logic in client applications is interesting, but I'd argue that libraries/SDKs are the ones that should take care of that. Very possible I'm missing something.

That said, building tree-like MCP applications does raise many questions. For example, imagine a leaf server requesting sampling from a root LLM through a few middle layers. It might help to have standard guidelines for forwarding requests and handling notifications, so each node doesn’t need to invent its own pass-through logic. Same goes for the other direction, i.e. notifying a downstream branch.

It's hard to speculate at this stage, we'll know more in a month or two when different patterns emerge.

**Comment by allenporter** [comment]
This sounds to me like a problem to solve at the discovery level. That is, say you have an server that exports a discovery endpoint like https://github.com/orgs/modelcontextprotocol/discussions/84 you can define separate MCP server endpoints for serving. That is, you can use http urls as the namespace.

I think you'll continue to run into a set of things that break when trying to treat separate MCP servers as a single server (e.g. oauth with a proxy of numerous MCP subservers with their own oauth endpoints for each subset of the server sounds complicated).  This may be fine for aggregating simple public servers -- but when trying to generalize further there will probably continue to be a number of things like this where trying to add more isolation within the MCP protocol will be complex for clients to adopt.

**Comment by patwhite** [comment]
@ggoodman I'm working on a similar problem, I think you're spot on with how you're thinking about the general flow here.

One question - could this all be as simple as introducing the notion of hierarchy to the various name fields? 

So, a prompt name like "code_review" becomes "@my-code-reviewer/code_review" or "my-code-reviewer.com/code_review"
and a tool call follows a similar pattern?
Then, for actual namespace scoping for general calls like tools/list, you could namespace before that:
"my-code-review.com/tools/list"
OR 
add a scoping property:
```
{
  "jsonrpc": "2.0",
  "id": 1,
  "method": "tools/list",
  "params": {
    "namespace": "my-code-reviewer.com",
    "cursor": "optional-cursor-value"
  }
}
```


Myself and @justinwilaby are working on a standardized search capability for servers with high volumes of tools and resources, and I think this fits in nicely with what we're working on:
https://github.com/modelcontextprotocol/specification/issues/204

We're specifically thinking about the enterprise use case of a proxy that proxies to numerous servers downstream from it and might have thousands or tens of thousands of tools exposed, way too many to simply list, and where you absolutely will have name collisions.

**Comment by mclenhard** [comment]
Hey @ggoodman ,

I had similar feelings and built something you might find helpful. It sort of takes the approach @jspahrsummers mentioned, where it acts as a server of servers. 

 It's essentially a central ingress/load balancer that can route traffic to different mcp servers based on things like tool call names. [link](https://github.com/mclenhard/catie-mcp) if you are curious. It's completely open-source; hope it helps with your use case.

**Comment by sterankin** [comment]
@mclenhard just to confirm.  Lets say you have many MCP servers (with different urls).  
You also have many different clients which could access those servers.  Can all clients be routed through some sort of gateway proxy rather than accessing each MCP server directly?

This proxy gateway would:
- handle auth, does the client have access to the MCP servers
- limit scope, clients should only 'see' the MCP servers that they have access to (maybe based on token claims?)
- be registry of all available MCP Servers and tools
- act as a proxy to MCP server calls, so clients only ever call one endpoint for tool usage

Does that make sense?  Has anyone built anything like this>?

**Reply by mclenhard** [reply]
> You also have many different clients which could access those servers. Can all clients be routed through some sort of gateway proxy rather than accessing each MCP server directly?

That is correct. As for your other points, three and four are fully handled. For OAuth, right now, it just forwards the request. It does maintain a cache of sessions to ensure that requests get forwarded to whatever server was previously responsible for that session, but there's definitely more work to be done here. 

Limiting scope is a good idea, and I've been thinking about it myself. I just haven't settled on the proper way to handle this yet. How much should go into my gateway vs. the application itself? I would love your thoughts if you have any here. Ideas and PR's are always welcome!

**Reply by mquadrat** [reply]
What you are describing is essentially an API gateway. 

Always keep in mind that we have two identities on the client side. The identity of the actual client application and the one of the user on whose behalf the calls are made.

**Reply by mclenhard** [reply]
Yeah, it's an API gateway. Good call out. I didn't even think about the identity of the client itself, just the user. Probably some interesting tooling to be built there as well.

**Reply by mquadrat** [reply]
It's the same set of problems like in API management. Seems natural that we use the same or at least similar solutions. Of course my opinion might be biased coming from the APIM side of things :D

**Reply by chhamilton** [reply]
Regarding client (not user) identity: there are many use cases that are enabled by this, including anti-abuse mechanisms, quality of service mechanisms, "peering" agreements, auditability, etc.

There's already significant momentum behind using [HTTP Message Signatures](https://www.rfc-editor.org/rfc/rfc9421.html) to identify Agents on the web, as well as other non-AI use cases (crawlers, aggregators, proxy operators). See this [recent blogpost by CloudFlare](https://blog.cloudflare.com/en-us/web-bot-auth/), which mentions OpenAI's adoption of this standard. I know similar conversations have been happening across other anti-abuse vendors and agent operators. There's also [some conversation](https://github.com/jhoyla/draft-jhoyla-bot-auth-use-cases/pull/1/commits/bab4879613452d17b72ba63ce3a4127d6e40ded5) around using "optiona mTLS" as an alternative mechanism (one is much more "web developer persona" friendly, the other is more "network admin persona" friendly).

Since MCP servers may be proxies or aggregators of traffic themselves, I think it's going to be important for the MCP protocol to add identity mechanisms internally. Both when initialization a connection (for receiving and validating the identity of the entity making the inbound request), and also for outbound requests. I can also imagine wanting to chain these identities such that downstream entities would be able to say "I'm receiving this request from MCP client X who received a request from Agent Y". Making the chain of operators verifiable would enable many downstream use cases.

It's probably in the interest of MCP folks to participate in these conversations, and also to adopt these standards. For this to work well I also think we're going to want any A2A protocols to participate as well.

**Comment by duaraghav8** [comment]
I'm currently solving the issue of namespacing by using slash separation.

For context, I'm building a [MCP registry+Proxy](https://github.com/duaraghav8/MCPJungle) (exactly what @sterankin describes above) focused on enterprises.

Every MCP server registered in my proxy is its own namespace. Tools can only be called by using the format `<server>/<tool name>`.
This allows different tools to have same name and still not cause a conflict.
eg- `github/git_commit` and `github2/git_commit` - same tools from 2 different MCP servers in my registry.

But I've already run into several issues here:
1. Claude outright [rejects](https://github.com/anthropics/claude-code/issues/985#issuecomment-2973621122) the tools from my proxy because of the `/` (its regexp doesn't allow slash in tool name)
2. Tool names get distorted in Cursor because it removes the `/` and it ends up showing names like `githubgit_commit`).

So it seems like currently Anthropic is very restrictive on the whole namespacing concept.

The only good solution I found is to use double underscore `__` as the separator instead of `/`.
But this seems too restrictive and unfriendly for developers.

Do people here have opinions on this? Would be great to have some guidance on how this should be handled properly

**Reply by LucaButBoring** [reply]
It looks like [a lot](https://github.com/search?q=repo%3Aanthropics%2Fclaude-code+%22%5E%5Ba-zA-Z0-9_-%5D%7B1%2C64%7D%24%22&type=issues) of people are running into issues with that regex. Having no spec-defined limitations on tool names (beyond "a string") means strict validations are entirely left up to client application implementers when limitations of some sort are needed for security or other reasons. Claude Code requiring alphanumeric+dash+underscore tool names is much more restrictive than the spec allows for, though.

Might be worth having a spec-defined limitation here for clients to use where necessary.

**Reply by duaraghav8** [reply]
I agree. Tool naming format should be defined as part of MCP spec.
I also just found [this issue](https://github.com/modelcontextprotocol/modelcontextprotocol/issues/277) that asks for it.

**Comment by olaservo** [comment]
Hi, sharing some related notes that seem relevant here from  the Technical Steering Committee meeting at the MCP Dev Summit on 5/23/25:

## More context from the discussion:

The suggestion below emerged during discussions about:
- MCP proxy servers that aggregate multiple upstream servers
- Security concerns about malicious servers duplicating tool names
- The need for logical grouping without breaking existing implementations

The committee consensus was that namespacing should be:
- A documented best practice, especially for proxy servers
- Not a mandatory protocol requirement
- Implementable without spec changes using current constraints

## Plan 9-inspired examples:

In [Plan 9](https://9p.io/plan9/), everything is a file and namespaces are per-process.

You can "mount" remote resources into your local namespace at any path, creating a unified hierarchical view.

How this concept could be applied to MCP Servers and Tools:

**Hierarchical Organization**

Just like Plan 9's file system, tools would have full paths:

```
/github/repos/search
/github/repos/create
/github/issues/list
/slack/channels/post
/slack/users/lookup
```

**Remote Server Mounting**

A proxy server could dynamically mount remote MCP servers, eg:

```python
# Conceptual example
proxy.mount("https://api.github.com/mcp", "/github/")
proxy.mount("https://slack-mcp.com", "/slack/")
proxy.mount("local-file-server", "/files/")
```

## Advantages mentioned at the meeting:

- **"achieving backward compatibility without reserving characters"** - Existing tools keep their names; the path prefix is added by the mounting layer
- **"can incorporate URLs of remote MCP servers"** - The mount point could be derived from or include the server's URL
- **"uses a string prefix, which is easier to enforce"** - Simple string concatenation rather than complex escaping rules
## Current allowed character limitations

As noted in the meeting (and by @duaraghav8 above), MCP tool names must match `^[a-zA-Z0-9_-]{1,64}$`, so actual forward slashes can't be used. Implementations would need to encode the hierarchy, eg:

```
/web1/search_web → web1_search_web
/github/repos/create → github_repos_create
```
## Other approaches discussed for avoiding Tool name collisions and naming:

- Simple server-based prefixing (e.g., `web1_search_web`) -> Covered in [this PR](https://github.com/modelcontextprotocol/modelcontextprotocol/pull/701)
- Generated/random prefixes (e.g., `jrwxs_search_web`) ->  Covered in [this PR](https://github.com/modelcontextprotocol/modelcontextprotocol/pull/701)
- Future additions like display names and labels/tags -> Covered in [this PR](https://github.com/modelcontextprotocol/modelcontextprotocol/pull/663/files) and future proposals

**Reply by patwhite** [reply]
@olaservo Thanks for writing this up - the main thing that this misses vs [this PR](https://github.com/modelcontextprotocol/modelcontextprotocol/pull/334) is the ability to scope down tool lists by namespace. So, while this handle the naming collision issue, it does NOT handle namespaces as a way to implement interfaces or "workspaces" where a small set of tools are composed for better grokability by agents (ie smaller sets of tools to select from rather than huge lists).

@evalstate did some tests that current LLM context max out remarkably early with tools, something in the range of 100 tools I believe (please correct me if I'm wrong). So, a major component of #334 is ways to scope the tool sets down. That DOES require a small protocol change, and I think is an incredibly important because it allows scoping down of lists. So, a client could conceivably get or search namespaces, or present namespaces to the user to select, then have a constrained set of tools.

A bit of information that I found really good - this isn't JUST for proxying, it's important for companies with large api surface areas. Microsoft's office api (Graph) has 26,000 endpoints - if 100 tool is the max, that's 260 servers they have to maintain, expose, and teach the models how to orchestrate across (I have no idea how they do that tbh), or teach their business analyst users how to connect to different servers. This is as opposed to, with namespaces, where you could namespace by feature area, then expose some targeted namespaces like "send email" that expose tools draft email, send email, spell check. 

@cliffhall  ^ you might like this use case too for similar reasons.

So, without a protocol change, you are still limited in the number of tools on a server because you can't list tools from just a single namespace, so we haven't really accomplished anything with "guidance."

**Reply by olaservo** [reply]
Thanks @patwhite, you're absolutely right that the mounting approach I described from the Dev Summit doesn't by itself address the tool discovery/listing problem for large servers.  Apologies for maybe combining topics too much here.

To clarify a little more: I think the 'guidance vs prescriptive' comment was specifically focused on the name collision aspect, not so much the scale/filtering challenge.

(Edited)

**Comment by scottpersinger** [comment]
My only small reservation here would be that the function names can impact LLM behavior, so a system where the function names can change easily might impact re-use of prompts or agent reliability. If user A mounts Slack endpoints at "slack/" and user B mounts them at "comms/" then you could get pretty divergent behavior. I suspect people end up promoting a "recommended" prefix anyway...

**Reply by LucaButBoring** [reply]
Part of this also comes down to having high-quality tool descriptions - if I mount Slack tools at `comms/` and the tool descriptions don't specify that those are Slack tools, the model probably won't know what I want if I ask it to post in a Slack channel, as opposed to if I mount those tools at `slack/` instead. But if the tool description does have that information, it shouldn't matter either way. There may be some divergence but I'd expect (hope?) it'd be negligible.

That puts the onus strongly on server authors to write good tool descriptions, which isn't always the case, however.

**Reply by richardkmichael** [reply]
I think this will also be less of an issue as clients hopefully gain more awareness of the server [`instructions`](https://modelcontextprotocol.io/specification/2025-06-18/basic/lifecycle) property in the [InitializeResult](https://github.com/modelcontextprotocol/typescript-sdk/blob/c40193933a43b61d40e616b7ab513e87b3d978de/src/types.ts#L348-L361).  The server can tell the client about itself, providing more context to the client.

I have tried to use this in my own MCP servers, but unfortunately, at least Claude seems unaware of it.  I haven't used other clients.

**Comment by zhjch05** [comment]
Sharing my project MetaMCP, a MCP proxy + aggregator. You can group MCP servers into namespaces then assign endpoints to them. You can also filter tools out, search feature WIP. https://github.com/metatool-ai/metamcp

Btw I think there are 2 dimensions logical group may consider: like typical blog posts, there are categories and tags concepts, where a blog can only fall into one category, but can have many tags, and the filter by tag can be both "AND" or "OR" logic.

Similarly from a tool search perspective, a category is a elasticsearch like "index", then within the index you can filter by tags, apply search dsls, or have vector search combined in hybrid mode to take in user intent from natural language.

**Reply by patwhite** [reply]
Ya, you and I are working on similar projects - if you check out #322 you'll see a full proposal around search, I'm 100% in agreement we need a protocol level mechanism for search

**Comment by luafran** [comment]
I don't see this explicitly mentioned here but In an MCP gateway, the tool, prompt, resource names are used for routing to the right MCP server implementing that specific resource. It seems unlikely to happen, but implementations should consider that a tool name may include the used separator string (like ___). MCP gateways can't just assume that the last part of the tool name after splitting using the separator is the actual resource name and anything else is the server name.

**Reply by duaraghav8** [reply]
I actually thought about this a lot. As of today, given the lack of naming convention in MCP spec and the strict RegExp in Claude code & desktop, the double/triple underscore is the best way to create separation between tool name & server name within a Gateway for routing.
The compromise here is then that such a Gateway should reject any tool name that itself contains `___`. IMO this is not a big issue because mostly nobody uses it in tool names, from what I've seen.

But I would definitely love to see:
1. Claude makes its tool naming RegExp a lot more permissive
2. MCP spec contains the regexp on tool naming (also permissive)

**Comment by kinthaiofficial** [comment]
Proxy clients that fan out to many backend servers is exactly the multi-tenant pattern that needs careful design. A few considerations:

**Budget allocation across backend servers.** When one MCP proxy serves many backend servers, the proxy needs to track which client is consuming which backend's resources. If Client A is hammering Backend Server B through the proxy, the proxy should throttle Client A's access to B — not throttle all clients, not throttle all backends.

**Capability aggregation vs. conflict resolution.** When two backend servers expose tools with the same name but different schemas, the proxy must decide: namespace them (prefix with server name), merge them (pick the more permissive schema), or conflict-reject. Naive aggregation creates subtle bugs when a client thinks it's calling Server A's "search" but actually calls Server B's.

**Trust propagation through the proxy.** If Backend Server A has a higher trust level than the client, can the proxy elevate the client's permissions for calls to A? It shouldn't. Trust should flow from the weakest link — if the client is community-signed, calls through the proxy to a platform-verified server should still be treated as community-signed.

**Session continuity across backend failures.** If Backend Server B goes down mid-workflow, the proxy should gracefully handle the failure — either by routing to a fallback server B2 that has the same capabilities, or by returning a partial result with a clear signal that B's capabilities are temporarily unavailable.

Related multi-tenancy isolation patterns: https://blog.kinthai.ai/openclaw-multi-tenancy-why-vm-per-user-doesnt-scale

**Comment by friendlygeorge** [comment]
Sharing some practical experience from running a Docker MCP server (50 tools, manages containers/services) as a de facto proxy pattern in production:

**1. Tool namespacing works but has a subtle failure mode.** We use `docker_` prefix for all tools (`docker_list_containers`, `docker_start`, etc.). This avoids name collisions cleanly. But: when a client calls a namespaced tool and the backend container is stopped, the error message contains the full namespaced name. LLMs sometimes retry with slight variations (`docker_start_container` vs `docker_start`), creating confusing retry loops. A standardized error format from the proxy would help clients distinguish "tool not found" from "backend unreachable."

**2. Graceful degradation matters more than perfect routing.** Our proxy pattern: if one backend container is down, the other 49 tools still work. We return a specific error code for "backend unavailable" vs "tool error" vs "auth failure." Clients that treat all errors the same way end up in degraded states. The spec could benefit from a standard error taxonomy for proxy servers.

**3. State is the hard part.** Each backend has its own state (container IDs, image tags, network configs). The proxy can not assume shared state. We solved this by making the proxy stateless — each tool call is independent, and the proxy routes to the right backend based on the tool name prefix. This works for infrastructure tools but would break for tools that need cross-backend transactions.

**4. Instructions concatenation is inadequate for heterogeneous backends.** As @ggoodman noted, naively concatenating instructions from 50 backends produces a wall of text that confuses models. We solved this by having the proxy maintain its own instructions that describe the *orchestration layer*, not individual backends. The proxy says "you have a Docker management interface with these capabilities" — not "here are 50 individual server descriptions."

**Comment by TimeToBuildBob** [comment]
The proxy/aggregation patterns here resonate with my experience building `gptme serve --mcp`, which exposes gptme's built-in tool ecosystem (bash, file ops, web fetch, browser, Python, etc.) as a single MCP server. A few practical observations that the thread hasn't fully covered:

**Stateful tools break naïve proxy routing**

Most proxy discussion here assumes stateless tools (search, fetch, transform). But some of the most useful tool servers maintain per-session state — gptme's `bash` tool keeps a persistent shell with working directory and env vars intact across calls; a code interpreter tool might hold a live Python kernel. When you aggregate these through a proxy, you can't just load-balance across upstream connections: you need client-session affinity, or you'll get a user's shell state routed to a different backend on the next call.

The MCP spec has a clean stateful abstraction via the `initialize` handshake, but proxy implementations often collapse this by reusing upstream connections. A spec clarification distinguishing "session-scoped" from "stateless" tool servers at the capability level would let proxies make correct routing decisions without having to infer statefulness from tool names or documentation.

**Tool count degrades reliability faster than expected**

From running gptme sessions with varying tool counts: 8–12 focused tools works reliably; 15–20 starts showing measurable decision noise (the model picks ambiguously-similar tools); 30+ produces noticeably unpredictable selection. This matches the GPT-4 behavior @neilmanvar described — it's a context-dilution effect, not model-specific.

For proxy aggregators this is a practical argument for tool filtering as a first-class feature rather than an optional nicety. A client requesting "file operation tools only" from a proxy aggregating 10 servers should see meaningfully better task completion than one taking the union of all 300+ tools. MCP currently has no mechanism for a client to express capability intent during `initialize`, which means proxies have to guess or expose everything.

**Instructions concatenation needs a compact mode**

When gptme exposes ~28 tools, each tool's description, examples, and constraints together fill a substantial chunk of the system prompt. An aggregator concatenating 5–10 such servers' `instructions` fields would overflow practical context budgets fast.

A reasonable spec extension: distinguish a brief instructions summary from the existing full `instructions`. The proxy presents summaries by default during initialize; the client can request full instructions for specific servers when it decides to specialize. This mirrors progressive disclosure — give the model enough context to route, then drill in when it commits to a tool domain. The pattern is well-understood in tool documentation systems; it just hasn't landed in MCP yet.
